Add webpacker compatibility with config switch and installation generator

sgara opened the pull request:
After reading a few issues and related discussions (see #5829, #5551 and #5048), and willing to switch to webpacker myself, I wanted to propose a first step toward moving to webpacker. This is of course open to discussion. The goal here is to be able to simple switch to loading active_admin assets from webpack instead of sprockets. Also introduced a gemfile to generate development app running webpacker although there is no js nor css at the moment. 

To me, the next step would be to create a npm package from sources present in this repo already. Note that a few changes may be required, I'm not sure we can have one set of assets compatible with both webpack and sprockets. But that can be discussed separately I guess

deivid-rodriguez commented:
Thanks @sgara! I haven't missed this, I'll have a closer look soon but it looks like a good first initial step in principle!

sgara commented:
thanks @deivid-rodriguez, no hurry. I'm already taking a look at the next step. 
Moving assets to `app/javascript`, bundling with rollup for still being able to import through sprocket without breaking existing apps and keeping only one set of source files seems to be the way to go. We'll check that in another PR ;)

deivid-rodriguez commented on lib/active_admin/namespace_settings.rb:
```diff
@@ -106,5 +106,8 @@ class NamespaceSettings < DynamicSettingsNode
     register :include_default_association_filters, true
+
+    # Switch between asset pipeline and webpacker assets
+    register :use_webpacker, false
```
Can we instead toggle this on through an environment variable? I really appreciate the approach of enabling this in small steps, but I don't want to expose this kind of setting to end users yet since this is just a work in progress. Maybe set `ENABLE_ACTIVEADMIN_WEBPACKER_SUPPORT=true` in `config/boot.rb` of the generated application, and then check that when choosing to either use `_pack`, or `_link` helpers?

sgara replied:
Hum it does make sense to keep it private while working on it. But in this case, I would instead merge my work in a different order given #5865, and instead of merging in the order I developed it, I would merge it this way: 
- rework js files to remove the use of `ActiveAdmin` global constant that isn't necessary
- setup npm packaging and rollup (if rollup works for you of course, tell me if there another bundling option you'd prefer). JS assets would be compiled when pushing to the repo, and the generated bundle then used by sprockets
- release the npm package including assets ready to be imported by webpack
- add that `use_webpacker` switch when everything else is ready

deivid-rodriguez replied:
Sure, sorry I didn't yet provide feedback in the other PR. Your plan sounds good to me :+1:.

sgara replied:
starting with rollup in fact, as sprockets-es6 doesn't handle `imports`, see #5868

deivid-rodriguez commented on gemfiles/rails_60_webpacker.gemfile:
```diff
@@ -0,0 +1,11 @@
+gem "activerecord-jdbcsqlite3-adapter", "~> 60.0.rc1", platform: :jruby
+gem "sqlite3", "~> 1.4", platform: :mri
+
+gem "webpacker", "~> 4.x"
```
I think we can use `4.0` here? Or do you need a specific `webpacker` pre-release for this?

deivid-rodriguez commented on spec/support/rails_template.rb:
```diff
@@ -79,6 +79,19 @@
+  append_file 'app/javascript/packs/active_admin.js', "import './active_admin.css';"
+  gsub_file 'config/webpacker.yml', /^.*extract_css.*$/, <<-YML
+  extract_css: true
+  YML
```
Use `gsub_file 'config/webpacker.yml', /^.*extract_css.*$/, "\n  extract_css: true"` for consistency with other `gsub_file` in this generator.

sgara commented:
Finally getting there!
@deivid-rodriguez @javierjulio PR updated. I added generators and templates to have a seamless testing process, installation instructions, ...
Looking forward to have your comments!

nyrf commented:
@sgara I got error 

```
undefined method `html_safe' for nil:NilClass
Did you mean?  html_safe?
```

<img width="842" alt="WX20191120-163650@2x" src="https://user-images.githubusercontent.com/299532/69222260-14d8e900-0bb4-11ea-90be-f47383c6dceb.png">

sgara commented:
Thanks for the feedback @nyrf I’ll look into it 👍

nyrf commented:
@sgara I think that because of style file wouldn't be generated in development environment, so the style file  seems like `http://localhost:3000/7de125c1-80c7-4e9c-a40a-4a5547107245`, i add ` if Rails.env.production?` and then it works for me.

```ruby
active_admin_application.stylesheets.each do |style, options|
  stylesheet_tag = active_admin_namespace.use_webpacker ? stylesheet_pack_tag(style, options) : stylesheet_link_tag(style, options)
  text_node(stylesheet_tag.html_safe) if Rails.env.production?
end
```

sgara commented:
@nyrf should be fixed now, thanks for testing 👍

nyrf commented:
@sgara It works. thanks for your work.

javierjulio commented on docs/0-installation.md:
````diff
@@ -112,6 +112,49 @@ Draper::CollectionDecorator.send :delegate, :per_page_kaminari
+    "window.jQuery":"jquery"
+  })
+)
+```
````
Since we have a webpacker generator now (thanks!) could we not include this there? There is an option to insert into an existing file or append to a file so this should be easy to automate. I'd suggest going off [some of the existing install files in webpacker](https://github.com/rails/webpacker/tree/master/lib/install) to get ideas. With that change then you'd also be able to remove this code from the support template too.

sgara replied:
Good point! PR updated accordingly

maokomioko commented:
Any chances we can merge it before 2.6 release? =^..^=

vfonic commented on lib/generators/active_admin/webpacker/templates/active_admin.js:
```diff
@@ -0,0 +1,19 @@
+// import 'active_admin/assets/stylesheets/active_admin'
```
These comments in this file seem to be misleading. Should we remove them all and just keep the actual code?

sgara replied:
Good point, comments cleaned/updated 👍

vfonic commented on lib/generators/active_admin/webpacker/webpacker_generator.rb:
```diff
@@ -0,0 +1,26 @@
+
+        insert_into_file Rails.root.join("config/webpack/environment.js").to_s,
+          "const jquery = require('./plugins/jquery')\n",
+          after: /require\(('|")@rails\/webpacker\1\);?\n/
```
Nice! :) Good covering of both cases. :) 👍

vfonic commented on spec/support/rails_template.rb:
```diff
@@ -58,7 +58,7 @@
 
 # Setup Active Admin
-generate 'active_admin:install'
+generate "active_admin:install#{" --use-webpacker" if ENV["BUNDLE_GEMFILE"] == File.expand_path("../../gemfiles/rails_60_webpacker/Gemfile", __dir__)}"
```
Should we switch to webpacker as default for rails 6+, just like the rails team did?

/cc @deivid-rodriguez 

This line would then be more like
```rb
generate "active_admin:install#{" --use-webpacker" if rails >= 6}" # Rails::VERSION::MAJOR or smth, don't know by heart
```

deivid-rodriguez replied:
We can consider that in the future, but initially I'd like this to be optin.

deivid-rodriguez commented:
Hi, so this is now on the top of my list, so I'll review and merge this very soon (I'm feeling a bit sick this week so that might delay it a bit, though). Anyways, so sorry for the delay here.

deivid-rodriguez commented on lib/generators/active_admin/install/install_generator.rb:
```diff
@@ -8,13 +8,15 @@ class InstallGenerator < ActiveRecord::Generators::Base
 
       hook_for :users, default: "devise", desc: "Admin user generator to run. Skip with --skip-users"
       class_option :skip_comments, type: :boolean, default: false, desc: "Skip installation of comments"
+      class_option :use_webpacker, type: :boolean, default: false, desc: "User webpacker assets instead of sprocket"
```
```suggestion
      class_option :use_webpacker, type: :boolean, default: false, desc: "User webpacker assets instead of sprockets"
```

deivid-rodriguez commented on lib/generators/active_admin/webpacker/webpacker_generator.rb:
```diff
@@ -0,0 +1,26 @@
+        template 'print.scss', 'app/javascript/packs/active_admin/print.scss'
+
+        copy_file "#{__dir__}/plugins/jquery.js", Rails.root.join("config/webpack/plugins/jquery.js").to_s
+
```
I think the last commit lost some stuff that's important to get tests green. At least I need something like

```diff
+        gsub_file Rails.root.join("config/webpacker.yml"), /^(.*)extract_css.*$/, '\1extract_css: true'
+
```

here to get some scenarios to pass (the ones about asset registration).

sgara replied:
Added for tests only as it's the only way to get `css` files and check for them to be present. Not necessary for development app

deivid-rodriguez commented on spec/support/rails_template.rb:
```diff
@@ -91,5 +91,14 @@
 
+# Add both assets to test as some specs still rely on sprockets
+generate "active_admin:assets"
+generate "active_admin:webpacker"
```
`generate "active_admin:webpacker"` shouldn't be necessary because that's done by `generate "active_admin:install --webpacker"`, right? At least I get some "identical" messages for the assets when I run the generator, meaning that the files are already there.

Regarding regenerating sprockets assets, I think we can remove the dependency on sprockets assets from our tests and kill that too.

I made the following changes to be able to do that:

```diff
diff --git a/spec/requests/stylesheets_spec.rb b/spec/requests/stylesheets_spec.rb
index 6e67cf047..f34922a5f 100644
--- a/spec/requests/stylesheets_spec.rb
+++ b/spec/requests/stylesheets_spec.rb
@@ -1,6 +1,6 @@
 require 'rails_helper'
 
-RSpec.describe "Stylesheets", type: :request do
+RSpec.describe "Stylesheets", type: :request, unless: ActiveAdmin.application.use_webpacker do
   require "sprockets"
 
   let(:css) do
diff --git a/spec/support/rails_template.rb b/spec/support/rails_template.rb
index 0bfa5953a..5b1a3b02d 100644
--- a/spec/support/rails_template.rb
+++ b/spec/support/rails_template.rb
@@ -1,7 +1,15 @@
 # Rails template to build the sample app for specs
 
-create_file 'app/assets/stylesheets/some-random-css.css'
-create_file 'app/assets/javascripts/some-random-js.js'
+webpacker_app = ENV["BUNDLE_GEMFILE"] == File.expand_path("../../gemfiles/rails_60_webpacker/Gemfile", __dir__)
+
+if webpacker_app
+  create_file 'app/javascript/packs/some-random-css.css'
+  create_file 'app/javascript/packs/some-random-js.js'
+else
+  create_file 'app/assets/stylesheets/some-random-css.css'
+  create_file 'app/assets/javascripts/some-random-js.js'
+end
+
 create_file 'app/assets/images/a/favicon.ico'
 
 require 'active_admin/dependency'
diff --git a/spec/unit/generators/install_spec.rb b/spec/unit/generators/install_spec.rb
index 0798dd31f..1362fa499 100644
--- a/spec/unit/generators/install_spec.rb
+++ b/spec/unit/generators/install_spec.rb
@@ -3,13 +3,21 @@ require 'rails_helper'
 RSpec.describe "AA installation" do
   context "should create" do
     it "active_admin.scss" do
-      path = Rails.root + "app/assets/stylesheets/active_admin.scss"
+      path = if ActiveAdmin.application.use_webpacker
+               Rails.root + "app/javascript/stylesheets/active_admin.scss"
+             else
+               Rails.root + "app/assets/stylesheets/active_admin.scss"
+             end
 
       expect(File.exist?(path)).to eq true
     end
 
     it "active_admin.js" do
-      path = Rails.root + "app/assets/javascripts/active_admin.js"
+      path = if ActiveAdmin.application.use_webpacker
+               Rails.root + "app/javascript/packs/active_admin.js"
+             else
+               Rails.root + "app/assets/javascripts/active_admin.js"
+             end
 
       expect(File.exist?(path)).to eq true
     end
```

With those changes, everything becomes green for me, except for 

```

Failures:

  1) ActiveAdmin::Views::Pages::Layout should be the default translation
     Failure/Error: stylesheet_tag = active_admin_namespace.use_webpacker ? stylesheet_pack_tag(style, options) : stylesheet_link_tag(style, options)
     
     Sprockets::Rails::Helper::AssetNotFound:
       The asset "active_admin.css" is not present in the asset pipeline.
     # ./lib/active_admin/views/pages/base.rb:32:in `block (2 levels) in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:31:in `each'
     # ./lib/active_admin/views/pages/base.rb:31:in `block in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:26:in `build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:8:in `build'
     # ./spec/unit/views/pages/layout_spec.rb:30:in `block (3 levels) in <top (required)>'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `new'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `arbre'
     # ./spec/support/active_admin_integration_spec_helper.rb:40:in `render_arbre_component'
     # ./spec/unit/views/pages/layout_spec.rb:29:in `block (2 levels) in <top (required)>'
     # ./spec/unit/views/pages/layout_spec.rb:43:in `block (2 levels) in <top (required)>'

  2) ActiveAdmin::Views::Pages::Layout should have lang attribute on the html element
     Failure/Error: stylesheet_tag = active_admin_namespace.use_webpacker ? stylesheet_pack_tag(style, options) : stylesheet_link_tag(style, options)
     
     Sprockets::Rails::Helper::AssetNotFound:
       The asset "active_admin.css" is not present in the asset pipeline.
     # ./lib/active_admin/views/pages/base.rb:32:in `block (2 levels) in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:31:in `each'
     # ./lib/active_admin/views/pages/base.rb:31:in `block in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:26:in `build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:8:in `build'
     # ./spec/unit/views/pages/layout_spec.rb:30:in `block (3 levels) in <top (required)>'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `new'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `arbre'
     # ./spec/support/active_admin_integration_spec_helper.rb:40:in `render_arbre_component'
     # ./spec/unit/views/pages/layout_spec.rb:29:in `block (2 levels) in <top (required)>'
     # ./spec/unit/views/pages/layout_spec.rb:47:in `block (2 levels) in <top (required)>'

  3) ActiveAdmin::Views::Pages::Layout should be the @page_title if assigned in the controller
     Failure/Error: stylesheet_tag = active_admin_namespace.use_webpacker ? stylesheet_pack_tag(style, options) : stylesheet_link_tag(style, options)
     
     Sprockets::Rails::Helper::AssetNotFound:
       The asset "active_admin.css" is not present in the asset pipeline.
     # ./lib/active_admin/views/pages/base.rb:32:in `block (2 levels) in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:31:in `each'
     # ./lib/active_admin/views/pages/base.rb:31:in `block in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:26:in `build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:8:in `build'
     # ./spec/unit/views/pages/layout_spec.rb:30:in `block (3 levels) in <top (required)>'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `new'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `arbre'
     # ./spec/support/active_admin_integration_spec_helper.rb:40:in `render_arbre_component'
     # ./spec/unit/views/pages/layout_spec.rb:29:in `block (2 levels) in <top (required)>'
     # ./spec/unit/views/pages/layout_spec.rb:37:in `block (2 levels) in <top (required)>'

  4) ActiveAdmin::Views::Pages::Layout the body should have namespace class
     Failure/Error: stylesheet_tag = active_admin_namespace.use_webpacker ? stylesheet_pack_tag(style, options) : stylesheet_link_tag(style, options)
     
     Sprockets::Rails::Helper::AssetNotFound:
       The asset "active_admin.css" is not present in the asset pipeline.
     # ./lib/active_admin/views/pages/base.rb:32:in `block (2 levels) in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:31:in `each'
     # ./lib/active_admin/views/pages/base.rb:31:in `block in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:26:in `build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:8:in `build'
     # ./spec/unit/views/pages/layout_spec.rb:30:in `block (3 levels) in <top (required)>'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `new'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `arbre'
     # ./spec/support/active_admin_integration_spec_helper.rb:40:in `render_arbre_component'
     # ./spec/unit/views/pages/layout_spec.rb:29:in `block (2 levels) in <top (required)>'
     # ./spec/unit/views/pages/layout_spec.rb:56:in `block (3 levels) in <top (required)>'

  5) ActiveAdmin::Views::Pages::Layout the body should have class 'active_admin'
     Failure/Error: stylesheet_tag = active_admin_namespace.use_webpacker ? stylesheet_pack_tag(style, options) : stylesheet_link_tag(style, options)
     
     Sprockets::Rails::Helper::AssetNotFound:
       The asset "active_admin.css" is not present in the asset pipeline.
     # ./lib/active_admin/views/pages/base.rb:32:in `block (2 levels) in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:31:in `each'
     # ./lib/active_admin/views/pages/base.rb:31:in `block in build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:26:in `build_active_admin_head'
     # ./lib/active_admin/views/pages/base.rb:8:in `build'
     # ./spec/unit/views/pages/layout_spec.rb:30:in `block (3 levels) in <top (required)>'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `new'
     # ./spec/support/active_admin_integration_spec_helper.rb:36:in `arbre'
     # ./spec/support/active_admin_integration_spec_helper.rb:40:in `render_arbre_component'
     # ./spec/unit/views/pages/layout_spec.rb:29:in `block (2 levels) in <top (required)>'
     # ./spec/unit/views/pages/layout_spec.rb:52:in `block (3 levels) in <top (required)>'

Finished in 12.15 seconds (files took 8.1 seconds to load)
120 examples, 5 failures

Failed examples:

rspec ./spec/unit/views/pages/layout_spec.rb:40 # ActiveAdmin::Views::Pages::Layout should be the default translation
rspec ./spec/unit/views/pages/layout_spec.rb:46 # ActiveAdmin::Views::Pages::Layout should have lang attribute on the html element
rspec ./spec/unit/views/pages/layout_spec.rb:34 # ActiveAdmin::Views::Pages::Layout should be the @page_title if assigned in the controller
rspec ./spec/unit/views/pages/layout_spec.rb:55 # ActiveAdmin::Views::Pages::Layout the body should have namespace class
rspec ./spec/unit/views/pages/layout_spec.rb:51 # ActiveAdmin::Views::Pages::Layout the body should have class 'active_admin'
```

Those failures I managed to fix them by moving the `use_webpacker` setting from `namespace_settings.rb` to `application_settings.rb` and using `active_admin_application.use_webpacker` instead of `active_admin_namespace.user_webpacker` when rendering the packs inside `lib/active_admin/views/pages/base.rb`.

But that changes how this PR works because right now `use_webpacker` is toggleable at the namespace level, and with those changes it would only be toggleable for the whole activeadmin application.

However, I think it should be possible to change the test appropriately to get it to pass with the current implementation.

sgara replied:
Indeed, it seems that I couldn't get that to work - as we were having sprockets related tests with webpacker app. Checking `use_webpacker` is a good option to me 👍 
With those changes, no need to generate both confs indeed 👌

sgara replied:
I slightly changed your suggestion to not test `register_stylesheet` things at all while testing webpacker. I actually doubt it would make any sense to do so anyway as a webpacker user would play with packs instead of registering extra stylesheets. 
I'd more likely deprecate `register_stylesheet` when using webpacker, what do you think?

deivid-rodriguez replied:
mmm, not sure but I think that grants a separate discussion for which I need to learn more about all this :sweat_smile:. But for now we're using `register_stylesheet` to register the `active_admin/prints.scss` stylesheet even when using webpacker, so I think it makes sense to test it?

sgara replied:
Forget it, you're right :)

sgara replied:
Ok I've managed to make it work while keeping the setting at namespace level. Thanks for the input here 👍

deivid-rodriguez commented on docs/0-installation.md:
```diff
@@ -112,6 +112,30 @@ Draper::CollectionDecorator.send :delegate, :per_page_kaminari
 
+## webpacker
+
+Since rails 6.0 webpacker has become the default asset generator. You can opt in using webpacker for ActiveAdmin assets as well by setting `use_webpacker` at install or manually.
```
I'm not a great prose writer, but I added a small suggestion that maybe reads a bit better?

```suggestion
Since rails 6.0, webpacker has become the default asset generator. You can opt into using webpacker for ActiveAdmin assets as well by setting `use_webpacker` at installation time or manually.
```

deivid-rodriguez commented:
@sgara I just noticed that you also need to modify CircleCI configuration to actually run the test suite against a webpacker application. Basically copy what we already do against turbolinks.

sgara commented:
I have to go back into it, it's been such a while now I forgot most of it.
I've rebased the branch and added circleci jobs (nice catch 😆 ), I'll check on your comments

deivid-rodriguez commented on spec/support/rails_template.rb:
```diff
@@ -57,8 +65,14 @@
+# Setup webpacker if necessary
+if webpacker_app
+  rake "webpacker:install"
+  gsub_file 'config/webpacker.yml', /^(.*)extract_css.*$/, '\1extract_css: true' if ENV['RAILS_ENV'] == 'test'
```
Does it make sense to include this line inside the installer? We could replace only the `test:` section in the `webpacker.yml` file.

sgara replied:
I don’t think we should force that on client app. We only do so to test that AA assets are present

deivid-rodriguez replied:
Ok, yeah, if this is not the default we should change that. You're right :+1:.

deivid-rodriguez commented on spec/support/rails_template.rb:
```diff
@@ -91,5 +100,10 @@
+if ENV["BUNDLE_GEMFILE"] == File.expand_path("../../gemfiles/rails_60_turbolinks/Gemfile", __dir__)
+  append_file 'app/assets/javascripts/active_admin.js', "//= require turbolinks\n"
+end
+
```
This is super picky, but nothing changed regarding `turbolinks`and still these lines have been shifted. Can we move them back to where they were to reduce and simplify the final diff?

deivid-rodriguez commented on spec/unit/views/pages/layout_spec.rb:
```diff
@@ -21,7 +21,11 @@
+    application = ActiveAdmin::Application.new
+    application.use_webpacker = ActiveAdmin.application.use_webpacker
+    ActiveAdmin::Namespace.new(ActiveAdmin.application, :myspace)
+  end
```
Nice! :)

sgara commented on spec/support/rails_template.rb:
```diff
@@ -57,8 +65,14 @@
 
+# Setup webpacker if necessary
+if webpacker_app
+  rake "webpacker:install"
```
We're settings things up for webpacker first here, as `--use-webpacker` expect the app to already have webpacker installed and set up (we need configuration files to be present)

deivid-rodriguez replied:
I see. Thanks!

deivid-rodriguez replied:
Could you use `rails_command` rather than `rake` for consistency with other rails stuff run by our generators?

javierjulio commented on docs/0-installation.md:
```diff
@@ -112,6 +112,30 @@ Draper::CollectionDecorator.send :delegate, :per_page_kaminari
 
+## webpacker
+
+Since rails 6.0, webpacker has become the default asset generator. You can opt into using webpacker for ActiveAdmin assets as well by setting `use_webpacker` at installation time or manually.
```
Just a slight difference since its technically new apps built with Rails 6 that get the default but not if you upgrade. I think this helps be a little clearer as to why its opt in for AA and would help avoid confusion.

```suggestion
For new apps starting with Rails 6.0, Webpacker has become the default asset generator. You can **opt-in to using Webpacker for ActiveAdmin assets** as well by updating your configuration to turn on the `use_webpacker` option, either at installation time or manually.
```

javierjulio commented on lib/generators/active_admin/install/install_generator.rb:
```diff
@@ -8,13 +8,15 @@ class InstallGenerator < ActiveRecord::Generators::Base
 
       hook_for :users, default: "devise", desc: "Admin user generator to run. Skip with --skip-users"
       class_option :skip_comments, type: :boolean, default: false, desc: "Skip installation of comments"
+      class_option :use_webpacker, type: :boolean, default: false, desc: "Use webpacker assets instead of sprockets"
```
```suggestion
      class_option :use_webpacker, type: :boolean, default: false, desc: "Use Webpacker assets instead of Sprockets"
```

javierjulio commented on lib/generators/active_admin/install/templates/active_admin.rb.erb:
```diff
@@ -325,4 +325,11 @@ ActiveAdmin.setup do |config|
+
+  # == Webpacker
+  #
+  # By default, Active Admin uses sprocket's asset pipeline.
```
```suggestion
  # By default, Active Admin uses Sprocket's asset pipeline.
```

javierjulio commented on lib/generators/active_admin/install/templates/active_admin.rb.erb:
```diff
@@ -325,4 +325,11 @@ ActiveAdmin.setup do |config|
+  # == Webpacker
+  #
+  # By default, Active Admin uses sprocket's asset pipeline.
+  # You can switch to webpacker here.
```
```suggestion
  # You can switch to using Webpacker here.
```

javierjulio commented on spec/requests/stylesheets_spec.rb:
```diff
@@ -1,6 +1,6 @@
 require 'rails_helper'
 
-RSpec.describe "Stylesheets", type: :request do
+RSpec.describe "Stylesheets", type: :request, unless: ActiveAdmin.application.use_webpacker do
```
Very cheeky. I didn't know we could skip a spec like this. 👌🏻

Any value in having a similar spec like this for webpacker, if we don't already? My guess is no but figured I'd ask.

sgara replied:
I'm pretty sure it would be too late to test for css errors in rails specs :) any error should appear when bundling with rollup anyway

javierjulio commented on spec/support/rails_template.rb:
```diff
@@ -57,8 +65,14 @@
 
+# Setup webpacker if necessary
+if webpacker_app
+  rake "webpacker:install"
```
I think its just a straight swap or other changes necessary @deivid-rodriguez?
```suggestion
  rails_command "webpacker:install"
```

javierjulio commented on CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@
 ### Enhancements
 
 * Extend menu to allow for nested submenus. [#5994] by [@taralbass]
+* Add webpacker compatibility with config switch and installation generator. [#5855] by [@sgara]
```
```suggestion
* Add Webpacker compatibility with opt-in config switch and installation generator. [#5855] by [@sgara]
```

sgara commented:
I’ll make the changes tonight I think. Glad to see the end of the road!

javierjulio commented on lib/generators/active_admin/webpacker/templates/active_admin.js:
```diff
@@ -0,0 +1,5 @@
+// Load Active Admin's styles into webpacker,
```
```suggestion
// Load Active Admin's styles into Webpacker,
```

sgara commented:
Applied changes from both @deivid-rodrigez and @javierjulio reviews and cleaned git history a bit.

deivid-rodriguez commented:
Great, I'll be releasing a new version very soon, thanks everyone, specifically @sgara for his massive effort!

ziaulrehman40 commented:
Why did we not let users use both webpacker and sprockets at the same time?
It is very possible and used widely as well for some use cases.

If i make a PR to fix this, would that be accepted?

javierjulio commented:
@ziaulrehman40 yes, I'd be interested to see what's involved.
